fix(transformer): use correct scope for TS namespaces

overlookmotel opened the pull request:
Re-use scope ID from `TSModuleDeclaration` for the scope of the function which replaces it in transformed output.

This should mean the scope tree is left in correct state after the transform.

NB: Still incomplete - we also need to transfer the binding ID from `X` in `namespace X {}` to `let X` in output.

overlookmotel commented:
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/3489?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#3489** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/3489?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 👈
* **#3488** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/3488?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @overlookmotel and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="11px" height="11px"/> <b>Graphite</b></a>
  <!-- Current dependencies on/for this PR: -->

overlookmotel commented:
Damn! Something in conformance is failing. I'll fix.

Dunqing commented on crates/oxc_transformer/src/typescript/namespace.rs:
```diff
@@ -133,36 +133,40 @@ impl<'a> TypeScript<'a> {
-        );
+        let real_name = match decl.id {
+            TSModuleDeclarationName::Identifier(ident) => ident.name,
+            TSModuleDeclarationName::StringLiteral(lit) => {
+                // TODO: Need to convert to valid identifier
+                lit.value
+            }
```
We don't care about `StringLiteral`, that only works for `module "x"`. Here we only transform `namespace`. You can check the TSModuleDeclaration's kind field. https://oxc-project.github.io/oxc/playground/?code=3YCAAIChgICAgICAgIC3mMl9fWP2DOEtS3aewAVF3MlN9R8zpdZMUnii0U5sQf%2Fxyrjgn39L%2BQCA

Dunqing replied:
But `module "d"` is not a valid typescript code. [Check this](https://www.typescriptlang.org/play/?target=1&lib=true&ts=5.5.0-dev.20240531#code/LYewJgrgNgpgBAIgM4LgbwL4CgthgYygEMAneUSWRFdDIA)

overlookmotel replied:
Babel test case `typescript/module-namespace/head/input.ts` includes this illegal syntax:

```ts
namespace N {}
namespace M.N.O {}
module M {}
module "m" {}
```

Should we be catching this and rejecting it in semantic?

overlookmotel replied:
I've skipped that Babel test case for transformer conformance, so we can assume `module "d"` is not possible here.

overlookmotel replied:
Have opened #3498 for catching this illegal syntax in semantic.

overlookmotel commented:
@boshen @Dunqing Can I ask for help here?

Conformance is failing for various TS compiler test cases due to panic on this line:

https://github.com/oxc-project/oxc/blob/1f1449f4fbb725b51b1ab7b20d4449805fb1b23b/crates/oxc_transformer/src/typescript/namespace.rs#L150

i.e. semantic has not set a `scope_id` for this `TSModuleDeclaration`.

Here's one that fails:

`typescript/tests/cases/compiler/declareDottedModuleName.ts`

```rs
// @declaration: true
module M {
    module P.Q { } // This shouldnt be emitted
}

module M {
    export module R.S { }  //This should be emitted
}

module T.U { // This needs to be emitted
}
```

The common factor among the failing tests appears to be the `@declaration: true` directive. When I remove that line, the tests pass. Does that directive disable semantic / change it's behavior somehow?

overlookmotel commented:
I assume the answer to my question is this:

https://github.com/oxc-project/oxc/blob/ff7e8c7448b3c781166b4634b06491b37fdc8535/crates/oxc_semantic/src/builder.rs#L141-L151

Semantic does not visit TS definition files.

Next questions:
1. Is that correct behavior?
2. What should transformed output be for a TS definition file? Pass it through unchanged? Output an empty file (the way all `declare` statements are stripped out of a `.ts` file)?

overlookmotel commented:
Thanks for coming back. But I'm seeing something different from Babel.

Here's the TS conformance test case `compiler/defaultDeclarationEmitShadowedNamedCorrectly.ts` run through Babel: [Babel REPL](https://babeljs.io/repl#?browsers=defaults%2C%20not%20ie%2011%2C%20not%20ie_mob%2011&build=&builtIns=false&corejs=3.21&spec=false&loose=false&code_lz=PTAEAEBMFMGMBsCGAnRAXAlgewHYC5Q1kBXaAKBAgDMN5odEBbaAtACwwGcA6NTsjIwAOWZGlAAqUIk6hmoKsiyNQAIm7B2XVQG4y0AB4ixoDDjTRkVRLGigAKhxwBzTgB4ACgBoHAPlAA3mSgIaBCBB56oYQE9noAvvpGouJUxDiwmLhyiADW0J4-AML2or4AFAYEAaA40ADuoOUAlAQloqAAZIFCSkKcEfGg8a0OTq6FoO3I_kHRyNBoxMg4tcTw8NKyiDgAnglkScbiZhZWNnYefbJBiUcpoDDW6-IIMrIAsrtFyiJ15oFgqFelh-hFrgdDMdakxoJwhBdQABlZSLcaA6J0cRfH7CXD0cQAXlAACYdKBKEi2IhIFhGjA3qgsjgfJwsKBOLtGAAjLCbLigXDwXagYicRDcuigABuGEQhDYdk40HgVAAtIJjkCQlCHrBcJxXgt0HZiYw8tBysxuE9EC9mgcgA&debug=false&forceAllTransforms=false&modules=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=true&presets=&prettier=false&targets=&version=7.24.6&externalPlugins=%40babel%2Fplugin-transform-typescript%407.24.6&assumptions=%7B%7D)

Or is it just that Babel's not understanding from the `@declaration: true` directive that it's a TS definition file?

In [TS playground](https://www.typescriptlang.org/play/?filetype=d.ts#code/PTAEAEBMFMGMBsCGAnRAXAlgewHYC5Q1kBXaAKBAgDMN5odEBbaAtACwwGcA6NTsjIwAOWZGlAAqUIk6hmoKsiyNQAIm7AMOIcTSqA3GWgAPEWNBa00ZFUSxooACoccAc04AeAAoAaJwD5QAG8yUDDQIQIvQ3DCAkdDAF8jU1FxKmIcWExcOUQAa2hvPwBhR1F-AApjAiDQHGgAd1BKgEoCMtFQADJgoSUhTijE0ET2pxd3YtBO5ECQ2ORoNGJkHHrieHhpWUQcAE8kshSzcUtrW3tQLwHZEOSTtNAYW03xBBlZAFl9kuURBo4cQLcL9LCDKK3I4mU71JjQThCOwOADKymWk2CoXCdHEPz+wlw9HEAF5QAAmfSgSgotiISBYZowD6oHI4PycLCgTj7RgAIyw2y4oFw8H2oGInEQfLooAAbhhEIQ2A5ONB4FQALSCU7YsIwp6wXCcd5LdAOMmMArQSrMbgvRBvVpHIA) it produces errors!

Dunqing commented:
> Or is it just that Babel's not understanding from the `@declaration: true` directive that it's a TS definition file?
> 
> In [TS playground](https://www.typescriptlang.org/play/?filetype=d.ts#code/PTAEAEBMFMGMBsCGAnRAXAlgewHYC5Q1kBXaAKBAgDMN5odEBbaAtACwwGcA6NTsjIwAOWZGlAAqUIk6hmoKsiyNQAIm7AMOIcTSqA3GWgAPEWNBa00ZFUSxooACoccAc04AeAAoAaJwD5QAG8yUDDQIQIvQ3DCAkdDAF8jU1FxKmIcWExcOUQAa2hvPwBhR1F-AApjAiDQHGgAd1BKgEoCMtFQADJgoSUhTijE0ET2pxd3YtBO5ECQ2ORoNGJkHHrieHhpWUQcAE8kshSzcUtrW3tQLwHZEOSTtNAYW03xBBlZAFl9kuURBo4cQLcL9LCDKK3I4mU71JjQThCOwOADKymWk2CoXCdHEPz+wlw9HEAF5QAAmfSgSgotiISBYZowD6oHI4PycLCgTj7RgAIyw2y4oFw8H2oGInEQfLooAAbhhEIQ2A5ONB4FQALSCU7YsIwp6wXCcd5LdAOMmMArQSrMbgvRBvVpHIA) it produces errors!

Yes, the directive only works in typescript. But I don't see any mention of documentation about `@declaration` in the typescript document. BTW our typescript coverage test has already handled this directive https://github.com/oxc-project/oxc/blob/6f71541db0da3da7da0f79e33aeb462789782a24/tasks/coverage/src/typescript.rs#L238

overlookmotel commented:
Conformance finally passes! Assuming that #3500 is correct, I think this is ready for a last review.

NB: I had to disable 3 conformance tests which contain illegal `module "m" {}` syntax (without `declare`).

overlookmotel commented:
I changed my mind. Now just skipping ambient modules, so was able to re-enable the failing tests containing this invalid syntax.

overlookmotel committed:
fix(transformer): use correct scope for TS namespaces (#3489)

Re-use scope ID from `TSModuleDeclaration` for the scope of the function which replaces it in transformed output.

This should mean the scope tree is left in correct state after the transform.

NB: Still incomplete - we also need to transfer the binding ID from `X` in `namespace X {}` to `let X` in output.
